FIX: Sparse matrix addition/subtraction eliminates explicit zeros

ananyashreyjain opened the pull request:
Fixes #9318 
Binary operations on Sparse matrices removes the explicit zeros. These changes preserve the explicit zeros in the output matrix.

ananyashreyjain commented:
Any comments on this PR ?

perimosocordiae commented:
Can you run an informal benchmark to show the impact on running time / memory usage? It seems like there shouldn't be too much impact in the no-explicit-zeros case, but I'd like to have some numbers.

ananyashreyjain commented:
Benchmark reveals that there is some impact on the running time.

   | before       |after              |  ratio  |                                                                              |
   |---------------|------------------|---------|---------------------------------------------------------------------------|
   | [a2ffe09a]| 20b3c18c]   |           |                                                                                           |
   | 1.78ms      | 2.27ms       |  1 .28 | sparse.Arithmetic.time_arithmetic('csr', 'AA', '__sub__')|
   |  4.56ms     | 5.76ms       | 1.26   | sparse.Arithmetic.time_arithmetic('csr', 'AB', 'multiply') |
   | 4.45ms      | 5.50ms       | 1.23   | sparse.Arithmetic.time_arithmetic('csr', 'BB', '__add__')|
   | 1.89ms      | 2.33ms       | 1.23   | sparse.Arithmetic.time_arithmetic('csr', 'AA', '__add__') |
   | 3.97ms      | 4.85ms       | 1.22   | sparse.Arithmetic.time_arithmetic('csr', 'BA', '__add__') |
   | 4.54ms      | 5.54ms       | 1.22   | sparse.Arithmetic.time_arithmetic('csr', 'BB', 'multiply') |
   | 3.98ms      | 4.85ms       | 1.22   | sparse.Arithmetic.time_arithmetic('csr', 'AB', '__sub__') |
   | 4.00ms      | 4.88ms       | 1.22   | sparse.Arithmetic.time_arithmetic('csr', 'BA', '__sub__') |
   | 3.96ms      | 4.82ms       | 1.22   | sparse.Arithmetic.time_arithmetic('csr', 'AB', '__add__') |
   | 4.57ms      | 5.53ms       | 1.21   | sparse.Arithmetic.time_arithmetic('csr', 'BA', 'multiply') |
   | 2.01ms      | 2.37ms       | 1.18   | sparse.Arithmetic.time_arithmetic('csr', 'AA', 'multiply') |
   | 4.66ms      | 5.49ms       | 1.18   | sparse.Arithmetic.time_arithmetic('csr', 'BB', '__sub__') |
   | 33.01ms    | 35.15ms     | 1.06   | sparse.Arithmetic.time_arithmetic('csr', 'BB', '__mul__') |

ananyashreyjain commented:
@perimosocordiae  I tried to fix this but nothing could bring a significant change in running time . Do you have any ideas that can reduce the running time?

perimosocordiae commented:
My main concern is a performance hit for matrices without explicit zeros, which I expect is the most common case, and isn't impacted by the issue this PR fixes.

One approach might be to maintain two separate functions: the current one that ignores zeros, and a new one that preserves them. You could dispatch to the appropriate function based on known properties of the matrices, or even try checking for explicit zeros before dispatching.

That said, it may also be possible to improve the algorithm for preserving explicit zeros to make the overhead negligible in the no-zeros case, but I haven't looked into it very deeply.

ananyashreyjain commented:
Benchmark for the new commit : 

| before        |   after           |ratio|                                                                                           |
|----------------|-----------------|-------|---------------------------------------------------------------------------|
|  [a2ffe09a]  | [750fe675]  |        |                                                                                          |
|    1.91ms    |  2.17ms      |1.13 | sparse.Arithmetic.time_arithmetic('csr', 'AA', '__add__')|
|    1.97ms    | 2.19ms       |1.11 | sparse.Arithmetic.time_arithmetic('csr', 'AA', 'multiply')|
|    1.79ms    | 1.98ms      |1.10  |sparse.Arithmetic.time_arithmetic('csr', 'AA', '__sub__')|
|    4.03ms    | 4.36ms      |1.08  |sparse.Arithmetic.time_arithmetic('csr', 'BA', '__sub__')|
|    4.53ms    | 4.87ms      |1.07  |sparse.Arithmetic.time_arithmetic('csr', 'BB', '__add__')|
|    4.57ms    |4.86ms      |1.06  |sparse.Arithmetic.time_arithmetic('csr', 'BA', 'multiply')|
|   14.77ms   |15.63ms     | 1.06 |sparse.Arithmetic.time_arithmetic('csr', 'AB', '__mul__')|
|    4.11ms    |4.33ms      |1.05  |sparse.Arithmetic.time_arithmetic('csr', 'BA', '__add__')|
|    4.71ms    |4.17ms      |0.89  |sparse.Arithmetic.time_arithmetic('csr', 'AB', 'multiply')|
|    5.59ms    | 4.89ms      |0.88 | sparse.Arithmetic.time_arithmetic('csr', 'BB', 'multiply')|
|   44.59ms   | 35.17ms     | 0.79  |sparse.Arithmetic.time_arithmetic('csr', 'BB', '__mul__')|

ananyashreyjain commented:
@perimosocordiae  `csr_binop_csr_canonical()` is designed to not include explicit zeros so it checks the uncommon non-zero elements of both matrices so as to not include them in the final matrix . I removed those checks which may be the cause of benchmark results getting better for some cases.

ananyashreyjain commented:
Benchmark :

|    before   |  after             |ratio|                                                                                           |
|--------------|------------------|-------|---------------------------------------------------------------------------|
| [a2ffe09a]|  [6929a0ff]    |        |                                                                                          |
|    1.93ms |    2.05ms      |1.06 | sparse.Arithmetic.time_arithmetic('csr', 'AA', '__sub__')|
|    2.09ms |    2.20ms      |1.05  |sparse.Arithmetic.time_arithmetic('csr', 'AA', '__add__')|
|    4.50ms |    4.27ms      |0.95  |sparse.Arithmetic.time_arithmetic('csr', 'BA', '__sub__')|
|    5.12ms |    4.79ms      |0.94  |sparse.Arithmetic.time_arithmetic('csr', 'BB', '__add__')|
|    5.37ms |    4.89ms      |0.91  |sparse.Arithmetic.time_arithmetic('csr', 'AB', 'multiply')  |
|   18.55ms|    16.84ms    | 0.91  |sparse.Arithmetic.time_arithmetic('csr', 'BA', '__mul__')|
|  17.31ms |   15.34ms     | 0.89 | sparse.Arithmetic.time_arithmetic('csr', 'AB', '__mul__')|

ananyashreyjain commented:
Benchmark:
 | before    | after       |ratio|                                                                                                |
|--------------|-------------|------|-------------------------------------------------------------------------------|
  |[95029ad1]| [24433a94]|                                                                                                |
  | 2.01ms     |2.27ms      |1.13  sparse.Arithmetic.time_arithmetic('csr', 'AA', '__add__')|
   | 2.05ms     |2.29ms      |1.12  sparse.Arithmetic.time_arithmetic('csr', 'AA', 'multiply')|
   | 1.85ms     |2.03ms      |1.10  sparse.Arithmetic.time_arithmetic('csr', 'AA', '__sub__')|
   |4.18ms     |4.53ms      |1.08  sparse.Arithmetic.time_arithmetic('csr', 'BA', '__sub__')|
   |15.22ms    |16.11ms    |  1.06  sparse.Arithmetic.time_arithmetic('csr', 'AB', '__mul__')|
    |5.31ms     |5.02ms      |0.95  sparse.Arithmetic.time_arithmetic('csr', 'BB', 'multiply')|
   |38.28ms    |35.40ms     | 0.92  sparse.Arithmetic.time_arithmetic('csr', 'BB', '__mul__')|
    |4.78ms     |4.34ms     | 0.91  sparse.Arithmetic.time_arithmetic('csr', 'AB', 'multiply')|

@perimosocordiae reminder for review.

perimosocordiae commented on scipy/sparse/sparsetools/csr.h:
```diff
@@ -745,7 +745,7 @@ void csr_binop_csr_general(const I n_row, const I n_col,
             T result = op(A_row[head], B_row[head]);
 
-            if(result != 0){
+            if(result != 0 || A_row[head] == 0){
```
Why only check `A_row[head]` here, and not `B_row[head]`?

ananyashreyjain replied:
As far as I could understand its working only those indices are checked for which there is a non-zero value (or explicitly defined zero) in at least one matrix so if result is zero and one matrix has zero value then other should also have zero value at that index (and one of the matrix must have a explicit zero at that index).

ananyashreyjain commented:
Initially `result` was  added to `Cx` if it was not zero because we don't want zero values in `Cx` but this also removed the explicitly defined zero values . To include those zeros we need to check  `Ax` and `Bx` for zeros (Ax and Bx can have zeros only when they are explicitly defined by user) so in the case when matrix B has no non-zero values at the indices for which A has some non-zero values then the values in A  at those indices were included in the final matrix if they were not zero but Ax cannot have zero values unless they are explicitly defines by user , so I removed the condition which was checking `Ax` for zeros. Same goes for `Bx`. In the case when there is a non-zero value in both the matrices  at some index and value of `result` is zero we only need to check  the value in one matrix because if value in one matrix is zero then value in other matrix  at that index should be zero in case of addition and subtraction. 
I extended similar logic for `csr_binop_csr_general()` but I couldn't fully understand its working so I am little doubtful for the non-canonical matrices.

ananyashreyjain commented:
@perimosocordiae  I hope the explanation made some sense.

ananyashreyjain commented:
@perimosocordiae any comments?

ananyashreyjain commented:
@pv any comments?

perimosocordiae commented:
It's been a while with no other comments, so I've merged this PR. Thanks for the contribution and your patience, @ananyashreyjain !

ananyashreyjain commented:
Thanks @perimosocordiae !

rgommers commented:
Nice. Could one of you make that addition to the release notes (https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.3.0), under "backwards incompatible changes" or "other changes"?

perimosocordiae commented:
Done. Let me know if you'd prefer more detail, or a link to this PR.

rgommers commented:
thanks, that is good as is I think

jnothman commented:
FYI, this change is the cause of test failures in Scikit-learn (https://travis-ci.org/scikit-learn/scikit-learn/builds/504653347), where we had relied on the removal of zeros in multiplication. More specifically, we assumed that multiplication of CSR matrices would not create explicit zeros where one of the operands was sparsely/implicitly zero. Perhaps this was not a reasonable assumption, and we could develop a workaround.

perimosocordiae commented:
I'll follow up on the scikit-learn issue.

perimosocordiae commented:
@ananyashreyjain I'm going to revert this PR for now, so we avoid breaking downstream projects. I think we can get this PR back into shape and re-merge it, though.

Let's be sure to add some test cases, too, to prevent this kind of issue from re-appearing.

ananyashreyjain commented:
@perimosocordiae right now my midterm exams are going on so I won't get the time to fix this but I will start working on it as soon as they are over (i.e 14th march) .

perimosocordiae commented:
@ananyashreyjain no worries, and good luck with your exams. When you get back to this, make a new PR and reference this one. I added a basic test case that will prevent the same issue from re-appearing next time.
